cluster/src/main/java/org/apache/iotdb/cluster/log/manage/serializable/SyncLogDequeSerializer.java

```diff
@@ -0,0 +1,310 @@
+        }
+        // removed log, skip
+        else {
+          readLog(logReader, true);
+        }
+        count++;
+      }
+      logReader.close();
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+
+    return result;
+  }
+
+  // read single log
+  private Log readLog(FileInputStream logReader, boolean canSkip) throws IOException {
+    int logSize = ReadWriteIOUtils.readInt(logReader);
+    int totalSize = Integer.BYTES + logSize;
+
+    if (canSkip) {
+      logReader.skip(logSize);
+      removedLogSize += totalSize;
+      return null;
+    }
+
+    Log log = null;
+
+    try {
+      log = parser.parse(ByteBuffer.wrap(ReadWriteIOUtils.readBytes(logReader, logSize)));
+    } catch (UnknownLogTypeException e) {
+      e.printStackTrace();
+    }
+
+    logSizeDeque.addLast(totalSize);
+
+    return log;
+  }
+
+  @Override
+  public LogManagerMeta recoverMeta() {
+    if (meta == null && metaFile.exists() && metaFile.length() > 0) {
+      try {
+        FileInputStream metaReader = new FileInputStream(metaFile);
+        firstLogPosition = ReadWriteIOUtils.readLong(metaReader);
+        meta = LogManagerMeta.deserialize(
+            ByteBuffer.wrap(ReadWriteIOUtils.readBytesWithSelfDescriptionLength(metaReader)));
+        metaReader.close();
+      } catch (IOException e) {
+        e.printStackTrace();
+      }
+    }
+
+    return meta;
+  }
+
+  @Override
+  public void serializeMeta(LogManagerMeta meta) {
+    try {
+      metaOutputStream.getChannel().truncate(0);
+      ReadWriteIOUtils.write(firstLogPosition, metaOutputStream);
+      ReadWriteIOUtils.write(meta.serialize(), metaOutputStream);
+
+      this.meta = meta;
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+
+  }
+
+  @Override
+  public void close() {
+    try {
+      logOutputStream.close();
+      metaOutputStream.close();
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  // actually delete removed logs, this may take lots of times
+  // TODO : use an async method to delete file
+  private void deleteRemovedLog() {
+    try {
+      FileInputStream reader = new FileInputStream(logFile);
+      // skip removed file
+      for (int i = 0; i < firstLogPosition; i++) {
+        readLog(reader, true);
+      }
+
+      // begin to write
+      File tempLogFile = SystemFileFactory.INSTANCE
+          .getFile(
+              IoTDBDescriptor.getInstance().getConfig().getSystemDir() + File.separator + "raftLog"
+                  + File.separator + "logData.temp");
+
+      logOutputStream.close();
+      logOutputStream = new FileOutputStream(tempLogFile);
+      int blockSize = 4096;
+      long curPosition = reader.getChannel().position();
+      while (curPosition < logFile.length()) {
+        long size = Math.min(blockSize, logFile.length() - curPosition);
+        logOutputStream.write(ReadWriteIOUtils.readBytes(reader, (int) size));
+        curPosition = reader.getChannel().position();
+      }
```

Comment:
There is a much easier way to do this, it is `FileChannel.transferTo()`, you may have a look.

Reply:
Fixed~